core/src/heaviest_subtree_fork_choice.rs

```diff
@@ -604,61 +978,89 @@ impl HeaviestSubtreeForkChoice {
                         return None;
                     }
                 }
-                let heaviest_slot_on_same_voted_fork = heaviest_slot_on_same_voted_fork.unwrap();
+                let heaviest_slot_hash_on_same_voted_fork =
+                    heaviest_slot_hash_on_same_voted_fork.unwrap();
 
-                if heaviest_slot_on_same_voted_fork == last_voted_slot {
+                if heaviest_slot_hash_on_same_voted_fork == last_voted_slot_hash {
                     None
                 } else {
-                    Some(heaviest_slot_on_same_voted_fork)
+                    Some(heaviest_slot_hash_on_same_voted_fork)
                 }
             })
             .unwrap_or(None)
     }
 
     #[cfg(test)]
-    fn set_stake_voted_at(&mut self, slot: Slot, stake_voted_at: u64) {
-        self.fork_infos.get_mut(&slot).unwrap().stake_voted_at = stake_voted_at;
+    fn set_stake_voted_at(&mut self, slot_hash_key: SlotHashKey, stake_voted_at: u64) {
+        self.fork_infos
+            .get_mut(&slot_hash_key)
+            .unwrap()
+            .stake_voted_at = stake_voted_at;
     }
 
     #[cfg(test)]
-    fn is_leaf(&self, slot: Slot) -> bool {
-        self.fork_infos.get(&slot).unwrap().children.is_empty()
+    fn is_leaf(&self, slot_hash_key: SlotHashKey) -> bool {
+        self.fork_infos
+            .get(&slot_hash_key)
+            .unwrap()
+            .children
+            .is_empty()
     }
 }
 
 impl TreeDiff for HeaviestSubtreeForkChoice {
-    fn contains_slot(&self, slot: Slot) -> bool {
-        self.fork_infos.contains_key(&slot)
+    type TreeKey = SlotHashKey;
+    fn contains_slot(&self, slot_hash_key: &SlotHashKey) -> bool {
+        self.fork_infos.contains_key(slot_hash_key)
     }
 
-    fn children(&self, slot: Slot) -> Option<&[Slot]> {
+    fn children(&self, slot_hash_key: &SlotHashKey) -> Option<&[SlotHashKey]> {
         self.fork_infos
-            .get(&slot)
+            .get(&slot_hash_key)
             .map(|fork_info| &fork_info.children[..])
     }
 }
 
 impl ForkChoice for HeaviestSubtreeForkChoice {
+    type ForkChoiceKey = SlotHashKey;
     fn compute_bank_stats(
         &mut self,
         bank: &Bank,
         _tower: &Tower,
-        _progress: &mut ProgressMap,
+        progress: &mut ProgressMap,
         computed_bank_state: &ComputedBankState,
     ) {
         let ComputedBankState { pubkey_votes, .. } = computed_bank_state;
-
+        let mut start = Measure::start("compute_bank_stats_time");
         // Update `heaviest_subtree_fork_choice` to find the best fork to build on
-        let best_overall_slot = self.add_votes(
-            &pubkey_votes,
+        let root = self.root.0;
+        let (best_overall_slot, best_overall_hash) = self.add_votes(
+            pubkey_votes.iter().filter_map(|(pubkey, slot)| {
+                if *slot >= root {
```

Comment:
nice catch, how about moving the root check inside `generate_update_operations`?

Reply:
moved to a `if new_vote_slot < self.root {` check